source/particles/property_pool.cc

```diff
@@ -30,53 +31,97 @@ namespace Particles
 
   PropertyPool::~PropertyPool()
   {
-    Assert(currently_open_handles.size() == 0,
-           ExcMessage("This property pool currently still holds " +
-                      std::to_string(currently_open_handles.size()) +
-                      " open handles to memory that was allocated "
-                      "via allocate_properties_array() but that has "
-                      "not been returned via deallocate_properties_array()."));
+    if (n_properties > 0)
+      {
+        const unsigned int n_open_handles =
+          properties.size() / n_properties - currently_available_handles.size();
+        (void)n_open_handles;
+        Assert(n_open_handles == 0,
+               ExcMessage("This property pool currently still holds " +
+                          std::to_string(n_open_handles) +
+                          " open handles to memory that was allocated "
+                          "via allocate_properties_array() but that has "
+                          "not been returned via "
+                          "deallocate_properties_array()."));
+      }
   }
 
 
 
-  PropertyPool::Handle
-  PropertyPool::allocate_properties_array()
+  void
+  PropertyPool::reinit()
   {
-    PropertyPool::Handle handle = PropertyPool::invalid_handle;
     if (n_properties > 0)
       {
-        handle = new double[n_properties];
-        currently_open_handles.insert(handle);
+        const unsigned int n_open_handles =
+          properties.size() / n_properties - currently_available_handles.size();
+        (void)n_open_handles;
+        AssertThrow(n_open_handles == 0,
+                    ExcMessage("This property pool currently still holds " +
+                               std::to_string(n_open_handles) +
+                               " open handles to memory that was allocated "
+                               "via allocate_properties_array() but that has "
+                               "not been returned via "
+                               "deallocate_properties_array()."));
       }
 
-    return handle;
+    // Clear vectors and ensure deallocation of memory
+    std::vector<double>().swap(properties);
+    std::vector<Handle>().swap(currently_available_handles);
   }
 
 
 
-  void
-  PropertyPool::deallocate_properties_array(Handle handle)
+  PropertyPool::Handle
+  PropertyPool::allocate_properties_array()
   {
-    Assert(currently_open_handles.count(handle) == 1, ExcInternalError());
-    currently_open_handles.erase(handle);
-    delete[] handle;
+    PropertyPool::Handle handle = PropertyPool::invalid_handle;
+    if (n_properties > 0)
+      {
+        if (currently_available_handles.size() > 0)
+          {
+            handle = currently_available_handles.back();
+            currently_available_handles.pop_back();
+          }
+        else
+          {
+            handle = properties.size() / n_properties;
+            properties.resize(properties.size() + n_properties);
```

Comment:
These additional elements are default initialized, which in the best of cases might be a zero. Could we set the to `signaling_nan()`?

Reply:
Hmm, I would prefer to enforce initialization to 0.0. Would you be ok with this?
```
            properties.resize(properties.size() + n_properties, 0.0);
```

Reply:
Yes, go for it!